cohttp-mirage/src/client.ml

```diff
@@ -18,40 +18,58 @@
  *)
 
 open Lwt.Infix
-module Channel = Mirage_channel.Make (Conduit_mirage.Flow)
-module HTTP_IO = Io.Make (Channel)
 
-module Net_IO = struct
-  module IO = HTTP_IO
+module Make
+    (P : Mirage_clock.PCLOCK)
+    (R : Resolver_mirage.S)
+    (S : Conduit_mirage.S) =
+struct
+  module Channel = Mirage_channel.Make (S.Flow)
+  module HTTP_IO = Io.Make (Channel)
+  module Endpoint = Conduit_mirage.Endpoint (P)
 
-  type ctx = { resolver : Resolver_lwt.t; conduit : Conduit_mirage.t }
+  module Net_IO = struct
+    module IO = HTTP_IO
 
-  let sexp_of_ctx { resolver; _ } = Resolver_lwt.sexp_of_t resolver
+    type ctx = {
+      resolver : R.t;
+      conduit : S.t option;
+      authenticator : X509.Authenticator.t option;
+    }
 
-  let default_ctx =
-    { resolver = Resolver_mirage.localhost; conduit = Conduit_mirage.empty }
+    let sexp_of_ctx { resolver; _ } = R.sexp_of_t resolver
 
-  let connect_uri ~ctx:{ resolver; conduit } uri =
-    Resolver_lwt.resolve_uri ~uri resolver >>= fun endp ->
-    Conduit_mirage.client endp >>= fun client ->
-    Conduit_mirage.connect conduit client >>= fun flow ->
-    let ch = Channel.create flow in
-    Lwt.return (flow, ch, ch)
+    let default_ctx =
+      { resolver = R.localhost; conduit = None; authenticator = None }
 
-  let close_in _ = ()
-  let close_out _ = ()
+    let connect_uri ~ctx:{ resolver; conduit; authenticator } uri =
+      R.resolve_uri ~uri resolver >>= fun endp ->
+      Endpoint.client ?tls_authenticator:authenticator endp >>= fun client ->
+      match conduit with
+      | None -> failwith "conduit not initialised"
```

Comment:
Is it ok to use failwith instead of Lwt.raise or Lwt.failwith?

Reply:
`Lwt.fail` does not record the backtraceI believe